tests/functional/codegen/test_abi_decode.py

```diff
@@ -0,0 +1,463 @@
+    a, b = _abi_decode(Foo(addr).get_counter(), (uint256, String[5]), unwrap_tuple=False)
+    return a, b
+    """
+
+    c2 = get_contract(contract_2)
+
+    assert tuple(c2.foo(c.address)) == (1, "hello")
+
+
+def test_abi_decode_private_dynarray(get_contract, abi_encode):
+    code = """
+bytez: DynArray[uint256, 3]
+
+@internal
+def _foo(bs: Bytes[160]):
+    self.bytez = _abi_decode(bs, DynArray[uint256, 3])
+
+@external
+def foo(bs: Bytes[160]) -> (uint256, DynArray[uint256, 3]):
+    dont_clobber_me: uint256 = MAX_UINT256
+    self._foo(bs)
+    return dont_clobber_me, self.bytez
+    """
+    c = get_contract(code)
+    bs = [1, 2, 3]
+    encoded = abi_encode("(uint256[])", (bs,))
+    assert c.foo(encoded) == [2 ** 256 - 1, bs]
+
+
+def test_abi_decode_private_nested_dynarray(get_contract, abi_encode):
+    code = """
+bytez: DynArray[DynArray[DynArray[uint256, 3], 3], 3]
+
+@internal
+def _foo(bs: Bytes[1696]):
+    self.bytez = _abi_decode(bs, DynArray[DynArray[DynArray[uint256, 3], 3], 3])
+
+@external
+def foo(bs: Bytes[1696]) -> (uint256, DynArray[DynArray[DynArray[uint256, 3], 3], 3]):
+    dont_clobber_me: uint256 = MAX_UINT256
+    self._foo(bs)
+    return dont_clobber_me, self.bytez
+    """
+    c = get_contract(code)
+    bs = [
+        [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
+        [[10, 11, 12], [13, 14, 15], [16, 17, 18]],
+        [[19, 20, 21], [22, 23, 24], [25, 26, 27]],
+    ]
+    encoded = abi_encode("(uint256[][][])", (bs,))
+    assert c.foo(encoded) == [2 ** 256 - 1, bs]
+
+
+def test_abi_decode_return(get_contract, abi_encode):
+    contract = """
+@external
+def abi_decode(x: Bytes[64]) -> (address, int128):
+    return _abi_decode(x, (address, int128))
+    """
+
+    c = get_contract(contract)
+
+    encoded = abi_encode("(address,int128)", (TEST_ADDR, 123))
+
+    assert tuple(c.abi_decode(encoded)) == (EXPECTED_TEST_ADDR, 123)
+
+
+def test_abi_decode_annassign(get_contract, abi_encode):
+    contract = """
+@external
+def abi_decode(x: Bytes[32]) -> uint256:
+    a: uint256 = _abi_decode(x, uint256, unwrap_tuple=False)
+    return a
+    """
+    c = get_contract(contract)
+
+    encoded = abi_encode("uint256", 123)
+    assert c.abi_decode(encoded) == 123
+
+
+@pytest.mark.parametrize(
+    "input",
+    [
+        b"",  # Length of byte array is below minimum size of output type
+        b"\x01" * 96,  # Length of byte array is beyond size bound of output type
+    ],
+)
+def test_clamper(get_contract, assert_tx_failed, abi_encode, input):
+    contract = """
+@external
+def abi_decode(x: Bytes[64]) -> (uint256, uint256):
+    a: uint256 = empty(uint256)
+    b: uint256 = empty(uint256)
+    a, b = _abi_decode(x, (uint256, uint256))
+    return a, b
+    """
+    c = get_contract(contract)
+    assert_tx_failed(lambda: c.abi_decode(input))
+
+
+@pytest.mark.parametrize(
+    "len,output_typ,input",
+    [
+        (160, "DynArray[uint256, 3]", b""),
+        (160, "DynArray[uint256, 3]", b"\x01" * 192),
```

Comment:
hmm, i think this fails because of the input arg clamper, not because of the clamper in abi_decode

Reply:
but maybe we can't actually hit the upper bound clamper?

Reply:
oh, and because the typechecker prevents the input arg from exceeding the upper bound as well?

Reply:
right, because we already check `bytearray_len <= 160` when we deserialize the function args